### PR TITLE
feat(ivy): enhance [style] and [class] bindings to be animation aware

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

File: packages/core/src/render3/styling/player_factory.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {BindingType, Player, PlayerFactory} from '../interfaces/player';
+
+export function createPlayerFactory<T>(
```
Comment (mhevery) [inline packages/core/src/render3/styling/player_factory.ts]: This is public interface, needs detail docs

Comment (mhevery) [reply packages/core/src/render3/styling/player_factory.ts]: Wondering if `bindPlayerFactory` would be a better name. You are not actually creating a player here, just wrapping it up in a bundle. Also maybe change `PlayerFactory` to `BoundPlayerFactory`

File: packages/core/src/render3/interfaces/player.ts [hunk 2]
```diff
@@ -19,6 +19,17 @@ export interface Player {
...
 }
 
+export interface PlayerFactory<T> {
+  fn:
+      (element: HTMLElement, type: BindingType, values: {[key: string]: any},
+       currentPlayer: Player|null) => Player | null;
+  value: T;
+}
+
+export interface PlayerBuilder extends BindingStore {
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/player.ts]: Needs public doc. What is the difference between `PlayerBuilder` and `BindingStore`. Are these public docs? Can this be simplified? And how is the `PlayerBuilder` and `PlayerFactory` related?

File: packages/core/src/render3/players.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import './ng_dev_mode';
+
+import {getContext} from './context_discovery';
+import {getRootContext} from './discovery_utils';
+import {scheduleTick} from './instructions';
+import {ComponentInstance, DirectiveInstance, Player} from './interfaces/player';
+import {RootContextFlags} from './interfaces/view';
+import {EMPTY_ARR, addPlayerInternal, getOrCreatePlayerContext, getPlayerContext, getPlayersInternal, getStylingContext, throwInvalidRefError} from './styling/util';
+
+export function addPlayer(
```
Comment (mhevery) [inline packages/core/src/render3/players.ts]: Needs public doc

File: packages/core/src/render3/players.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  const element = context.native as HTMLElement;
+  const lViewData = context.lViewData;
+  const playerContext = getOrCreatePlayerContext(element, context) !;
+  const rootContext = getRootContext(lViewData);
+  addPlayerInternal(playerContext, rootContext, element, player, 0, ref);
+  scheduleTick(rootContext, RootContextFlags.FlushPlayers);
+}
+
+export function getPlayers(ref: ComponentInstance | DirectiveInstance | HTMLElement): Player[] {
```
Comment (mhevery) [inline packages/core/src/render3/players.ts]: Needs public doc

**Review by mhevery — CHANGES_REQUESTED** [review]
All public APIs need to have docs. (Not clear which methods are intended to be public)

Please document the methods. It is not clear what they do. The comments should say what it does, how it does it, and why its done that way (or what problem does it solve).

What I am missing (and I assume that others will to) is how all of these pieces fit together. Could we get together and discuss this? Perhaps create a animations.md file explaining how these things work and how they are intended to be used.

File: packages/core/src/render3/interfaces/player.ts (hunk 2 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/player.ts]: This is private

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**